### PR TITLE
Add bare bones example app

### PR DESCRIPTION
**Pull request by bradjc**
This example shows how to make the smallest app:

- It requests no memory.
- It does not use crt0.c.
- It just immediately calls yield.

### CODE REVIEW

File: examples/tests/bare_bones/main.c [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+            void* mem_start __attribute__((unused)),
+            void* memory_len __attribute__((unused)),
+            void* app_heap_break __attribute__((unused))) {
+  // Yield.
+  __asm__ volatile (
+      "svc 0       \n"
+      :
+      :
+      : "memory", "r0", "r1", "r2", "r3", "r12", "lr"
+      );
```
Comment (alistair23) [inline examples/tests/bare_bones/main.c:16]: This will fail to build for RISC-V

**Review by alistair23 — CHANGES_REQUESTED** [review]
Do you have an idea of what this would be useful for?

**Comment by lschuermann** [comment]
@alistair23 I think this is pretty awesome to have a template and starting point for developing an app which does not use `libtock-c`'s `crt0` bootstrapping, but shows you the way the kernel will perform the initial call into the app. I can imagine this can be rather useful if one were to port runtimes / interpreters / JITs to Tock, such as MicroPython (which might not want to use the regular `libtock-c` app skeleton).

**Comment by bradjc** [comment]
Good starting point if you want to skip crt0.c, and might be useful to test any assumptions the kernel is making about processes.

File: examples/tests/bare_bones/main.c [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#elif defined(__riscv)
+  __asm__ volatile (
+      "li    a0, 0\n"
```
Comment (lschuermann) [inline examples/tests/bare_bones/main.c:21]: This doesn't look right. AFAIK the system call type is passed in `a4` on RISC-V, with `a0` to `a3` being the arguments (where for yield we're probably going to want to load a well-defined value into `a0` anyways).
```suggestion
      "li    a0, 1\n" // yield-wait
      "li    a4, 0\n" // yield system call class
```

**Review by lschuermann — CHANGES_REQUESTED** [review]

File: examples/tests/bare_bones/main.c [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+void _start(void* app_start, void* mem_start, void* memory_len, void* app_heap_break);
+
+void _start(void* app_start __attribute__((unused)),
+            void* mem_start __attribute__((unused)),
+            void* memory_len __attribute__((unused)),
+            void* app_heap_break __attribute__((unused))) {
+  // Yield.
+#if defined(__thumb__)
+  __asm__ volatile (
+      "svc 0       \n"
```
Comment (lschuermann) [inline examples/tests/bare_bones/main.c:13]: We might want to deterministically execute a yield-wait call here as well. (https://github.com/tock/tock/blob/a66504cd9bc339d85e857bec0afd9797e8d81297/kernel/src/syscall.rs#L42-L48)

I don't really know ARM ASM so don't want to make a suggestion. :)

Comment (ppannuto) [reply examples/tests/bare_bones/main.c:13]: Adding a `movs r1, #1` before the `svc` should be portable across all cortex-m

File: examples/tests/bare_bones/main.c (hunk 2 above)
Comment (bradjc) [reply examples/tests/bare_bones/main.c:21]: Ohhhh, ok I'm not crazy. a0 was the correct register when I wrote this app.

File: examples/tests/bare_bones/main.c (hunk 3 above)
Comment (bradjc) [inline examples/tests/bare_bones/main.c:13]:
```suggestion
      "movs r0, #1 \n"
      "svc 0       \n"
```

**Review by alistair23 — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]

**Comment by bradjc** [comment]
Can we merge this?

**Comment by hudson-ayers** [comment]
I had seen the CI didn't pass and ignored it as a result

bors r+

**Pull request merged**